### DISCUSSION TITLE
Usage of gradient for functions that accept single argument

### DISCUSSION THREAD

**Discussion by sreeramcoder** [themetaschemer/malt#45] [Q&A] [answered]
Hello,

This is regarding frames 78:17 and 78:18.

As per the text, gradient function is designed to work exclusively with lists.

I see that the gradient accepts scalar as second argument for unary functions (e.g. sqr). 

Here is the snippet and the output.

```

> (require malt/learner) 

>  (gradient-of (lambda (x) (sqr x)) (list 10.0))                                                                                                                           
; vector-length: contract violation                                                                                                                                        
;   expected: vector?                                                                                                                                                      
;   given: '(#(#<procedure:dual> 10.0 #<procedure:end-of-chain>))                                                                                                          
; [,bt for context]      

                                                                                                                                                  
>  (gradient-of (lambda (x) (sqr x)) 10.0)                                                                                                                                  
20.0                                                                                                                                                                       

```
The example in frame 78:17 seems to contain a typo.

Below is in text (note the usage of theta-0 inside lambda, whereas lambda accepts theta as parameter)

`(gradient-of (lambda (theta) (sqr theta-0)) 10.0)         `

Instead it has to be below 
`
(gradient-of (lambda (theta) (sqr theta)) 10.0)`         

Regards,
Sreeram

**Comment by themetaschemer** [comment, marked as answer]
Sreeram, sqr function is not defined to work on lists, so (sqr x) in your example will not work. 

Technically, `gradient-of` takes a function of one argument and an argument and returns the gradient of that function at that argument. So when we have more than one parameter with which we need gradients, we package them in a list. 

So, in your first example you should say: 
```
(gradient-of (lambda (xl) (sqr (ref xl 0))) (list 10.0))
```
(this is also what the example in frame 78:17 does).

**Reply by sreeramcoder** [reply]
Thanks for your reply. This helps !
